[clang] Prevent duplicated instantiation of enumerators of unscoped member enumerations

thebrandre opened the pull request:
This commit addresses a bug occuring when an unscoped member enumeration of
a class template is introduced with an opaque-enum-declaration and later
redeclared with an enum-specifier (per C++23 [class.mem] p6).
Previously, the enumerators, or EnumConstantDecl, of the enum-specifier were
instantiated at both declarations, leading to different issues:
* erroneous ambiguities when referencing the enumerators,
* duplicated diagnostics in the enumerator-list.

The issue is resolved by ensuring that enumerators are instantiated only at
the first instantiated declaration, analogous to nested classes.

Fixes #124405

thebrandre committed:
[clang] Add tests with member enum redeclarations

thebrandre committed:
[clang] Fix issues on template class enum member redeclaration

thebrandre commented:
Here are some details why I think this is a valid fix:

If you take a look at the generated AST you can see that the EnumConstantDecl for `X 'S<int>::E'` appears twice in the instantiation in the AST for the following example.

```cpp
template <typename T>
struct S {
    enum E : T;
    enum E : T { X = 5 };
};

auto x = S<char>::X;
```

```
`-ClassTemplateSpecializationDecl <line:1:1, line:5:1> line:2:8 struct S definition implicit_instantiation
|   |-DefinitionData [...]
|   |-TemplateArgument type 'int'
|   | `-BuiltinType 'int'
|   |-CXXRecordDecl <col:1, col:8> col:8 implicit struct S
|   |-EnumDecl <line:3:5, col:14> line:4:10 E 'int'
|   | `-EnumConstantDecl <col:18, col:22> col:18 X 'S<int>::E'
|   |   `-ConstantExpr <col:22> 'int'
|   |     |-value: Int 5
|   |     `-IntegerLiteral <col:22> 'int' 5
|   `-EnumDecl prev 0x1b3d5d90 <col:5, col:14> col:10 E 'int'
|     `-EnumConstantDecl <col:18, col:22> col:18 X 'S<int>::E'
|       `-ConstantExpr <col:22> 'int'
|         |-value: Int 5
|         `-IntegerLiteral <col:22> 'int' 5
```

If you compare it with analog cases with member classes and scoped member enumerations such as the following ...

```cpp
template <typename T>
struct S {
    struct C;
    struct C {
        static constexpr T X = 5;
    };
};

auto x = S<int>::C::X;
```

```cpp
template <typename T>
struct S {
    enum class E : T;
    enum class E : T { X = 5 };
};

auto x = S<int>::E::X;
```
... you can see [here on Compiler Explorer](https://godbolt.org/z/TEh95eqcv) that the second declarations in the instantiation simply refer to the previous and do not generate another definition. 

Note that this issue does also not occur for function-local enumerations:

```cpp
template <typename T>
T f() {
    enum E : T;
    enum E : T { X = 5 };
    return (T)X;
}

auto x = f<int>();
```

See [here on Compiler Explorer](https://godbolt.org/z/djoW8soEn).

I also added tests to make sure that out-of-class definitions are still supported.

thebrandre commented:
@cor3ntin I came across this while refactoring, which is still in progress because I'm coming across odd things like this that make me a bit uncertain. Sometimes it's hard to tell if its a bug or a feature.

thebrandre committed:
[clang] Update release notes

thebrandre commented:
I almost forgot to mention that this is actually valid code.
[C++23 [class.mem] p6](https://wg21.link/class.mem#general-6)
> A member shall not be declared twice in the *member-specification* except that
> * a nested class or member class template can be declared and then later defined, and
> * an enumeration can be introduced with an *opaque-enum-declaration* and later redeclared with an *enum-specifier*.

cor3ntin reviewed (approved):
LGTM, thanks.
However can you add details to the github description (which ultimately becomes the git commit message)? thanks

thebrandre commented:
@cor3ntin Thanks! I just added some details to the GitHub description. I hope it's not too verbose.
And I rebased the branch to resolve the conflict in the release notes.
